File: 10.syntax-scoring/py/part2.py

import sys

scores = []
for line in sys.stdin:
    error = False
    stack = []
    for c in line.strip():
        if c in {'(', '[', '{', '<'}:
            stack.append(c)
        elif stack[-1] == {')': '(', ']': '[', '}': '{', '>': '<'}[c]:
            stack.pop()
        else:
            error = True
            break
    value = 0
    if not error:
        while len(stack) != 0:
            value = 5 * value + {'(': 1, '[': 2, '{': 3, '<': 4}[stack.pop()]
        scores.append(value)

print("Result:", sorted(scores)[len(scores) // 2])

File: 10.syntax-scoring/cpp/part2.cpp

#include <bits/stdc++.h>

using namespace std;

int main() {
    vector<long> line_scores;
    string sequence;
    while (cin >> sequence) {
        vector<char> stack;
        bool error = false;
        for (char c : sequence) {
            if (c == '(' || c == '[' || c == '{' || c == '<') {
                stack.push_back(c);
            } else {
                char exp = stack.back();
                stack.pop_back();
                if (
                    (exp == '(' && c != ')') || (exp == '[' && c != ']') || (exp == '{' && c != '}')
                    || (exp == '<' && c != '>')
                ) {
                    error = true;
                    break;
                }
            }
        }
        if (!error) {
            long score = 0;
            while (!stack.empty()) {
                score *= 5;
                char exp = stack.back();
                stack.pop_back();
                if (exp == '(') {
                    score += 1;
                } else if (exp == '[') {
                    score += 2;
                } else if (exp == '{') {
                    score += 3;
                } else {
                    score += 4;
                }
            }
            line_scores.push_back(score);
        }
    }
    nth_element(line_scores.begin(), line_scores.begin() + line_scores.size() / 2, line_scores.end());
    cout << "Result: " << line_scores[line_scores.size() / 2] << endl;
    return 0;
}

